modules/ethereum-bsc/src/lib.rs

```diff
@@ -0,0 +1,292 @@
+
+			// ensure valid length
+			assert!(last_authority_set.len() / 2 < headers.len(), "Invalid headers size");
+
+			let last_checkpoint = FinalizedCheckpoint::get();
+			let checkpoint = &headers[0];
+
+			// get configuration
+			let cfg: BSCConfiguration = T::BSCConfiguration::get();
+
+			// ensure valid header number
+			// CHECKME it should be <= or == ?
+			assert!(last_checkpoint.number + cfg.epoch_length == checkpoint.number, "Ridiculous checkpoint header number");
+
+			// ensure first element is checkpoint block header
+			assert!(checkpoint.number % cfg.epoch_length == 0, "First element is not checkpoint");
+
+			// verify checkpoint
+			// basic checks
+			verification::contextless_checks(&cfg, checkpoint, &T::ChainTime::default()).map_err(|e|e.msg())?;
+			// check signer
+			let signer = utils::recover_creator(checkpoint).map_err(|e| e.msg())?;
+			ensure!(Self::contains(last_authority_set, signer), <Error::<T>>::InvalidSigner);
+
+
+			// extract new authority set from submitted checkpoint header
+			let new_authority_set = &utils::extract_authorities(checkpoint).map_err(|e| e.msg())?;
+
+			// log already signed signer
+			let mut recently = BTreeSet::<Address>::new();
+
+			for i in 1..headers.len() {
+				verification::contextless_checks(&cfg, &headers[i], &T::ChainTime::default()).map_err(|e|e.msg())?;
+				// check parent
+				verification::contextual_checks(&cfg, &headers[i], &headers[i-1]).map_err(|e|e.msg())?;
+				// who signed this header
+				let signer = utils::recover_creator(&headers[i]).map_err(|e| e.msg())?;
+				// signed must in last authority set
+				ensure!(Self::contains(last_authority_set, signer), <Error::<T>>::InvalidSigner);
+				// headers submitted must signed by different authority
+				ensure!(!recently.contains(&signer), <Error::<T>>::SignedRecently);
+				recently.insert(signer);
+
+				// enough proof to finalize new authority set
+				if recently.len() == last_authority_set.len()/2 {
+					// already have N/2 valid headers signed by different authority separately
+					// finalize new authroity set
+					FinalizedAuthority::put(new_authority_set);
+					FinalizedCheckpoint::put(checkpoint);
+					T::OnHeadersSubmitted::on_valid_authority_finalized(submitter, new_authority_set);
+					// skip the rest submitted headers
+					return Ok(());
+				}
+			}
+			T::OnHeadersSubmitted::on_invalid_headers_submitted(submitter);
+			// <Error::<T>>::HeadersNotEnough
+		}
+	}
+}
+
+decl_storage! {
+	trait Store for Module<T: Config> as Bridge {
+		/// Finalized authority set.
+		FinalizedAuthority get(fn finalized_authority) config(): Vec<Address>;
+		FinalizedCheckpoint get(fn finalized_checkpoint) config(): BSCHeader;
+	}
+	add_extra_genesis {
+		config(initial_header): BSCHeader;
+		build(|config| {
+			// TODO should we verify this header?
+			initialize_storage::<T>(
+				&config.initial_header,
+			);
+		})
+	}
+}
+
+/// Initialize storage.
+#[cfg(any(feature = "std", feature = "runtime-benchmarks"))]
```

Comment:
It seems that this is not only used in `runtime-benchmarks`, should we have this feature.